src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/Json/EnumNameHelpers.cs

```diff
@@ -0,0 +1,164 @@
+        // For example, the following enum will fail to build because both fields would resolve to "OK":
+        //
+        // enum Status {
+        //     STATUS_OK = 0;
+        //     OK = 1;
+        // }
+        //
+        // Tooling error message:
+        // ----------------------
+        // Enum name OK has the same name as STATUS_OK if you ignore case and strip out the enum name prefix (if any).
+        // (If you are using allow_alias, please assign the same number to each enum value name.)
+        //
+        // Just in case it does happen, map to the first value rather than error.
+        foreach (var item in nameMappings.GroupBy(m => m.RemoveEnumPrefixName).Select(g => KeyValuePair.Create(g.Key, g.First().OriginalName)))
+        {
+            removeEnumPrefixMapping[item.Key] = item.Value;
+        }
+
+        return new EnumMapping { WriteMapping = writeMapping, RemoveEnumPrefixMapping = removeEnumPrefixMapping };
+    }
+
+    // Remove the prefix from the specified value. Ignore case and underscores in the comparison.
+    private static string TryRemovePrefix(string prefix, string value)
+    {
+        var normalizedPrefix = prefix.Replace("_", string.Empty, StringComparison.Ordinal).ToLowerInvariant();
+
+        var prefixIndex = 0;
+        var valueIndex = 0;
+
+        while (prefixIndex < normalizedPrefix.Length && valueIndex < value.Length)
+        {
+            if (value[valueIndex] == '_')
+            {
+                valueIndex++;
+                continue;
+            }
+
+            if (char.ToLowerInvariant(value[valueIndex]) != normalizedPrefix[prefixIndex])
+            {
+                return value;
+            }
+
+            prefixIndex++;
+            valueIndex++;
+        }
+
+        if (prefixIndex < normalizedPrefix.Length)
+        {
+            return value;
+        }
+
+        while (valueIndex < value.Length && value[valueIndex] == '_')
+        {
+            valueIndex++;
+        }
+
+        return valueIndex == value.Length ? value : value.Substring(valueIndex);
+    }
+
+    private static string GetEnumValueName(string enumName, string valueName)
+    {
+        var result = TryRemovePrefix(enumName, valueName);
+
+        // Prefix name starting with a digit with an underscore to ensure it is a valid identifier.
+        return result.Length > 0 && char.IsDigit(result[0])
+            ? $"_{result}"
+            : result;
+    }
+
+    private sealed class EnumMapping
+    {
+        public required Dictionary<object, NameMapping> WriteMapping { get; init; }
+        public required Dictionary<string, string> RemoveEnumPrefixMapping { get; init; }
+    }
+
+    private sealed class NameMapping
+    {
+        public required object Value { get; init; }
+        public required string OriginalName { get; init; }
+        public required string RemoveEnumPrefixName { get; init; }
+    }
```

Comment:
Can we make the `EnumMapping` and `NameMapping` as record?

``` csharp
private sealed record EnumMapping(
    Dictionary<object, NameMapping> WriteMapping, 
    Dictionary<string, string> RemoveEnumPrefixMapping);

private sealed record NameMapping(
    object Value, 
    string OriginalName, 
    string RemoveEnumPrefixName);
```

Reply:
Yes. But classes work fine here. And records are bloated when doing AOT.